api/queue/work/model_service_deployment.go

```diff
@@ -107,11 +107,9 @@ func (depl *ModelServiceDeployment) Deploy(job *queue.Job) error {
 	}
 
 	defer func() {
-		deploymentCounter.WithLabelValues(model.Project.Name, model.Name, fmt.Sprint(endpoint.Status), fmt.Sprint(isRedeployment)).Inc()
+		deploymentCounter.WithLabelValues(model.Project.Name, model.Name, fmt.Sprint(deployment.Status), fmt.Sprint(isRedeployment)).Inc()
 
 		// record the deployment result
-		deployment.Status = endpoint.Status
-		deployment.Error = endpoint.Message
```

Comment:
If I understood correctly, this object `deployment` corresponds to the 'new' deployment that we're attempting to  deploy right? If so, how does the change above prevent a running 'old' deployment's `Status` from being changed to anything that is not pending/serving/running when a 'new' re-deployment attempt fails?

Reply:
-> If I understood correctly, this object deployment corresponds to the 'new' deployment that we're attempting to deploy right?

Correct.

-> If so, how does the change above prevent a running 'old' deployment's Status from being changed to anything that is not pending/serving/running when a 'new' re-deployment attempt fails?

If you refer to the old deployments as the previous rows in the deployment table, we will not modify them. Instead, if you are referring to the running old version_endpoints, we will not modify the status if it's already running or serving (ref: https://github.com/caraml-dev/merlin/blob/main/api/queue/work/model_service_deployment.go#L90-L93)

Reply:
Got it, thanks for the clarification!

> If you refer to the old deployments as the previous rows in the deployment table, we will not modify them.

I'm actually referring to this. Your explanation on the running old version_endpoints makes sense. It answers why the status in (1) remains as 'running' (if I'm understanding things correctly). What about the status of the old deployment marked in (2)? I still don't get how that is fixed because it should be 'running' right? Just like in (1). 😣 Sorry if I misunderstood something in the current workflow.

<img width="1369" alt="Screenshot 2023-10-23 at 12 33 10 PM" src="https://github.com/caraml-dev/merlin/assets/36802364/49125f61-085d-414a-b091-56093cd4abe4">

Reply:
Can you elaborate on how to reproduce your issue? From my end, the old deployment status remains Deployed when the newer deployment failed:

<img width="500" alt="Screen Shot 2023-10-23 at 11 43 18" src="https://github.com/caraml-dev/merlin/assets/8122852/26e3b815-6f82-4d05-90d9-10634f3bae32">

Reply:
Okay, I think misunderstood how the implementation worked so I couldn't understand how the bug came about. In any case, I get how the whole thing works now and I see why your fix manages to address that issue. Thanks a lot for following up on this thread Arief! :D Everything's clear now!